src/components/GitHubSync.tsx

```diff
@@ -0,0 +1,947 @@
+            const commentBlockEl = document.querySelector(
+              `.rm-block__input[id$="${commentBlockUid}"]`
+            ) as HTMLDivElement;
+            if (!commentBlockEl) return;
+            window.roamAlphaAPI.ui.setBlockFocusAndSelection({
+              location: {
+                "block-uid": commentBlockUid,
+                "window-id": getUids(commentBlockEl).windowId,
+              },
+              selection: {
+                start: 0,
+                end: 0,
+              },
+            });
+          }, 100);
+        }}
+      />
+      <Button
+        icon="import"
+        title="Check for New Comments"
+        minimal
+        small
+        outlined
+        style={{ paddingLeft: "1rem", paddingRight: "1rem" }}
+        className="transition"
+        loading={loadingComments}
+        onClick={async () => {
+          setLoadingComments(true);
+          const pageUid = getPageUidByBlockUid(commentsContainerUid);
+          await insertNewCommentsFromGitHub({ pageUid, extensionAPI });
+          setLoadingComments(false);
+        }}
+      />
+    </div>
+  );
+};
+const GitHubSyncPage = ({
+  commentsContainerUid,
+  extensionAPI,
+}: {
+  commentsContainerUid: string;
+  extensionAPI: OnloadArgs["extensionAPI"];
+}) => {
+  useEffect(() => {
+    const headerCommentObserver = createBlockObserver({
+      onBlockLoad: (b) => {
+        const { blockUid } = getUids(b);
+        if (blockUid === commentsContainerUid) {
+          if (b.hasAttribute("github-sync-comment-header")) return;
+          b.setAttribute("github-sync-comment-header", "true");
+          const containerDiv = document.createElement("div");
+          containerDiv.className = "inline-block ml-2";
+          containerDiv.onmousedown = (e) => e.stopPropagation();
+          b.append(containerDiv);
+          ReactDOM.render(
+            <CommentsContainerComponent
+              commentsContainerUid={commentsContainerUid}
+              extensionAPI={extensionAPI}
+            />,
+            containerDiv
+          );
+        }
+      },
+    });
+    const childCommentObserver = createBlockObserver({
+      onBlockLoad: (b) => {
+        const { blockUid } = getUids(b);
+        const parentUid = getParentUidByBlockUid(blockUid);
+        if (parentUid === commentsContainerUid) {
+          if (b.hasAttribute("github-sync-comment")) return;
+          b.setAttribute("github-sync-comment", "true");
+          const containerDiv = document.createElement("div");
+          containerDiv.className = "inline-block ml-2";
+          containerDiv.onmousedown = (e) => e.stopPropagation();
+          b.append(containerDiv);
+          ReactDOM.render(
+            <CommentsComponent blockUid={blockUid} />,
+            containerDiv
+          );
+        }
+      },
+    });
+
+    return () => {
+      headerCommentObserver.forEach((o) => o.disconnect());
+      childCommentObserver.forEach((o) => o.disconnect());
+    };
```

Comment:
adding a console.log here didn't fire.  not sure why or if that means these observers aren't being cleaned up.

Reply:
yea this all doens't look right, this component is not mounting at all. I would replace the

```
ReactDOM.render(
      <GitHubSyncPage
        commentsContainerUid={commentsContainerUid}
        extensionAPI={extensionAPI}
      />,
      containerDiv
    );
```

from above with just the contents of this `useEffect`. Actually the "observing" is happening by the time we get to `renderGitHubSyncPage` so I wouldn't even initiate new observers for both. I would render `CommentsContainerComponent` directly there, and have a separate root observer for `CommentsComponent`

Reply:
A few problems I'm running into trying to implement this:

**Observers Required**
Both observers are required to re-render the buttons when a user clicks into a block.

If we just render `CommentsContainerComponent` and remove the observer that renders it, if a user clicks into the block the buttons will be gone until the page reloads.  Maybe a necessary trade-off?  But I'd like to avoid that for the UX.

**RE: Root Observer**
The `CommentsComponent` renders to children of the `CommentsContainer`.  To know if it's a child we need the `CommentsContainer` `el` or `uid` which we get from a user defined query on page load.  We wouldn't have that at root.

I guess we could add an attribute on block creation/page load, but we'd run into the re-render issue still (a user clicks into the block, buttons are gone until page is reloaded).  

Or maybe we keep a running global cache of comment uids for the current page that clears on page unload ... we have to handle the sidebar windows too ... hmmm.

Reply:
I fixed the cleanup not firing by changing `return null` to `return <></>;`

Reply:
Here's a cache implementation: https://github.com/RoamJS/query-builder/compare/github-sync...github-sync-comment-uid-cache

Reply:
merged into this branch: https://github.com/RoamJS/query-builder/pull/254